os/drivers/input/ist415_update.c

```diff
@@ -0,0 +1,667 @@
+			return -EAGAIN;
+		}
+
+		touchvdbg("Version compare IC: %x(%x, %x), BIN: %x(%x, %x)\n", fw->cur.release_ver, fw->cur.core_ver, fw->cur.config_ver, fw->bin.release_ver, fw->bin.core_ver, fw->bin.config_ver);
+
+		/* If FW version is same, check FW checksum */
+		if ((fw->cur.core_ver == fw->bin.core_ver) && (fw->cur.config_ver == fw->bin.config_ver) && (fw->cur.release_ver == fw->bin.release_ver) && (fw->cur.test_ver == 0)) {
+			ret = ist415_i2c_read(dev, HCOM_GET_CRC, (uint8_t *)&chksum, 2);
+			if (ret || (chksum != fw->chksum)) {
+				touchdbg("Mismatch checksum IC: %x, Bin: %x (ret: %d)\n", chksum, fw->chksum, ret);
+				return -EAGAIN;
+			}
+		}
+
+		/* If the ver of binary is higher than ver of IC, FW update operate. */
+		if ((fw->cur.core_ver >= fw->bin.core_ver) && (fw->cur.config_ver >= fw->bin.config_ver) && (fw->cur.release_ver >= fw->bin.release_ver)) {
+			return OK;
+		}
+	}
+
+	return -EAGAIN;
+}
+
+/****************************************************************************
+ * Name: ist415_fw_update
+ ****************************************************************************/
+
+static int ist415_fw_update(struct ist415_dev_s *dev, const uint8_t *buf, int size)
+{
+	struct ist415_fw *fw = &dev->fw;
+	uint32_t core_ver = ist415_parse_ver(dev, FLAG_CORE, buf);
+	uint32_t config_ver = ist415_parse_ver(dev, FLAG_CONFIG, buf);
+	uint32_t release_ver = ist415_parse_ver(dev, FLAG_RELEASE, buf);
+	uint32_t test_ver = ist415_parse_ver(dev, FLAG_TEST, buf);
+	uint32_t chksum = 0;
+	int ret = 0;
+
+	touchvdbg("*** Firmware update ***\n");
+	touchvdbg("core: %x, config: %x, release: %x, test: %x(addr: 0x%x ~ 0x%x)\n", core_ver, config_ver, release_ver, test_ver, fw->index, fw->index + fw->size);
+
+	ret = ist415_isp_fw_update(dev, buf);
+	if (ret < 0) {
+		touchdbg("Fail to update firmware\n");
+		goto reset;
+	}
+
+	ret = ist415_i2c_read(dev, HCOM_GET_CRC, (uint8_t *)&chksum, 2);
+	if (ret < 0) {
+		touchdbg("Fail to read checksum\n");
+		goto reset;
+	}
+
+	if (chksum != fw->chksum) {
+		touchdbg("Mismatch checksum IC: %x, Bin: %x\n", chksum, fw->chksum);
+		goto reset;
+	}
+
+	ret = ist415_read_info(dev);
+	if (ret < 0) {
+		touchdbg("Fail to read info\n");
+		goto reset;
+	}
+
+	touchvdbg("Firmware update success\n");
+
+	return OK;
+
+reset:
+
+	ist415_reset(dev, false);
+
+	return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: ist415_check_fw
+ ****************************************************************************/
+
+int ist415_check_fw(struct ist415_dev_s *dev)
+{
+	struct firmware firmware;
+	struct ist415_fw *fw = &dev->fw;
+	int retry = 0;
+	int ret = 0;
+
+	ret = request_firmware(&firmware, FIRMWARE_PATH_WASHING_MACHINE, &dev);
+	if (ret < 0) {
+		touchdbg("Don't request firmware\n");
```

Comment:
firmware->data is not released as release_firmware not called here

Reply:
I added free in error case in request_firmware()